[wasm][crypto] RandomNumberGenerator mapped to Web Crypto getRandomValues

kjpou1 opened the pull request:
- Uses Web Crypto API [`getRandomValues`](https://www.w3.org/TR/WebCryptoAPI/#Crypto-method-getRandomValues) if available.

- Add javascript bridge implementation library to Native source tree.  …
   - Javascript checks for crypto interface and uses `crypto.getRandomValues`
   - Add api bridge call when building for emscripten browser.
   - If we couldn't find a proper implementation, as Math.random() is not suitable we will abort.
```

ABORT: no cryptographic support found getRandomValues. Consider polyfilling it if you want to use something insecure like Math.random(), e.g. put this in a --pre-js: var crypto = { getRandomValues: function(array) { for (var i = 0; i < array.length; i++) array[i] = (Math.random()*256)|0 } };

```

closes https://github.com/dotnet/runtime/issues/42727

Part of https://github.com/dotnet/runtime/issues/40074 and https://github.com/dotnet/runtime/issues/40074#issuecomment-684065169

kjpou1 committed:
[wasm][crypto] RandomNumberGenerator mapped to Web Crypto getRandomValues

- Uses Web Crypto API [`getRandomValues`](https://www.w3.org/TR/WebCryptoAPI/#Crypto-method-getRandomValues) if available.
- Falls back to `/dev/urandom` as default if the crypto library is missing.

steveisok commented:
Doesn't `/dev/urandom` result in a call to `Crypto.getRandomValues` via the emscripten implementation?

kjpou1 commented:
Looks like it does:  https://github.com/paulshapiro/emscripten/blob/master/src/library_fs.js#L1296

We can close this if we want to continue using the default implementation

jkotas commented on src/libraries/Native/Unix/System.Native/pal_random.c:
```diff
@@ -55,6 +55,10 @@ void SystemNative_GetNonCryptographicallySecureRandomBytes(uint8_t* buffer, int3
 }
 
+#ifdef __EMSCRIPTEN__
+int32_t mono_wasm_browser_crypto_getRandomValues(uint8_t* buffer, int32_t bufferLength);
```
Can the implementation be here instead of in mono runtime?

kjpou1 replied:
The code that inlines the javascript code does not generate correctly from the dotnet tree.

```
EM_JS(int32_t, mono_wasm_browser_crypto_getRandomValues, (uint8_t* buffer, int32_t bufferLength),
 {
... inline javascript code
}
```

Either way the javascript code will end up being in one of the javascript libraries that this will call as well as atjhe definition to define it. 

The 

```
int32_t mono_wasm_browser_crypto_getRandomValues(uint8_t* buffer, int32_t bufferLength);
```

Will end up being something like the following anyway:

```
EMSCRIPTEN_KEEPALIVE int32_t 
mono_wasm_browser_crypto_getRandomValues(uint8_t* buffer, int32_t bufferLength);
```

And the code actually moved into the javascript libraries.  

The implantation here is actually clearer in my opinion.  Either way let me know if one prefers the above instead.

jkotas replied:
> The code that inlines the javascript code does not generate correctly from the dotnet tree.

What is the problem? Should we make it work correctly instead?

The System.*.Native shims are meant to be standalone and not depend on the runtime. The shims have no other calls to the runtime today.

kjpou1 replied:
It is the javascript code is all.  It needs to be in the javascript libraries so I will move it to there and there will be no reference to the runtime but will be linked with the javascript code implementation itself.

kjpou1 replied:
Ok, referencing it that way will not work either.  Not sure how we will get it without referencing that way.

jkotas replied:
Both pal_random.c  and driver.c are C files. Why does the javascript work in driver.c and not in pal_random.c ?

Is it because of the build steps for each are different? If yes, can we make the build steps for pal_random.c same as the build steps for driver.c so that the javascript works in pal_random.c as well?

lewing replied:
> Ok, referencing it that way will not work either. Not sure how we will get it without the reference the way it is.

What exactly are you trying and how is it failing?

kjpou1 replied:
This no longer inlines the JS in driver.c.

GrabYourPitchforks commented:
> Falls back to `/dev/urandom` as default if the crypto library is missing.

Is this even possible? Is there any wasm host which does not provide a subtle crypto implementation? And if there is, why not force the host to polyfill as in https://github.com/dotnet/runtime/pull/42731?

steveisok commented:
> > Falls back to `/dev/urandom` as default if the crypto library is missing.
> 
> Is this even possible? Is there any wasm host which does not provide a subtle crypto implementation? And if there is, why not force the host to polyfill as in #42731?

This is just about the best way to call https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues.  It's the only API that isn't promised based and we can actually invoke.  

To be honest, this already works and unless I'm missing something, I would recommend we close the PR.

GrabYourPitchforks commented:
Let me rephrase the requirement:

`RandomNumberGenerator`'s built-in implementation __must__ call `Crypto.getRandomValues`. It would be better if it made this call directly, but it would also be acceptable to make this call indirectly, _assuming there's no fallback code path_. The emscripten implementation doesn't seem to fulfill this latter requirement.

jkotas commented on src/libraries/Native/Unix/System.Native/pal_random.c:
```diff
@@ -68,6 +72,18 @@ int32_t SystemNative_GetCryptographicallySecureRandomBytes(uint8_t* buffer, int3
+        else
+            return 0;
+    }
+#endif
```
This needs to be `#else` to meet the requirement stated by @GrabYourPitchforks

kjpou1 replied:
addressed in recent commits

lewing commented:
https://github.com/emscripten-core/emscripten/pull/12240 landed recently which will expose getentropy as a very thin wrapper over subtle crypto which is what the existing wasm /dev/random uses.  The code does fall back to a polyfill on node and fails if no polyfill is available.  I propose we stick with using /dev/random until we update emscripten then call getentropy from pal_random for wasm.  Does this satisfy all the requirements?

lewing commented:
> > Falls back to `/dev/urandom` as default if the crypto library is missing.
> 
> Is this even possible? Is there any wasm host which does not provide a subtle crypto implementation? And if there is, why not force the host to polyfill as in #42731?

This just works because emscripten falls back to node's crypro.randomBytes for /dev/random which is what the code here should do as well instead of falling back to the device.

lewing reviewed (changes requested):
This has several issues, it should be implemented in terms of calling [getentropy](https://man7.org/linux/man-pages/man3/getentropy.3.html) regardless of where the implementation lives.

kjpou1 committed:
Remove the emscripten interface code from the driver.c.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-crypto-getRandomValues

kjpou1 committed:
remove extraneous code comment

kjpou1 committed:
Move emscripten definition around.

kjpou1 committed:
Address review comment

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-crypto-getRandomValues

kjpou1 committed:
Add javascript bridge implementation library to Native source tree.

- Javascript checks for crypto interface and uses `crypto.getRandomValues`
- Add api bridge call when building for emscripten browser.
- separate out into browser subdirectory
- If we couldn't find a proper implementation, as Math.random() is not suitable we will abort.

```

ABORT: no cryptographic support found getRandomValues. Consider polyfilling it if you want to use something insecure like Math.random(), e.g. put this in a --pre-js: var crypto = { getRandomValues: function(array) { for (var i = 0; i < array.length; i++) array[i] = (Math.random()*256)|0 } };

```

kjpou1 committed:
Change tests to set random values of the buffer instead of return a single value.

kjpou1 committed:
Remove old test code

kjpou1 committed:
Remove testing code

kjpou1 committed:
Incorporate the PAL bridge layer into the `--js-library` build process

jkotas commented on src/libraries/Native/Unix/System.Native/browser/pal_random_browser.h:
```diff
@@ -0,0 +1,9 @@
+#include "pal_compiler.h"
+#include "pal_types.h"
+
+PALEXPORT int32_t dotnet_browser_entropy(uint8_t* buffer, int32_t bufferLength);
```
This should not be `PALEXPORT`. The method is not exported from the PAL.

kjpou1 replied:
very true.  addressed in recent commits

jkotas commented on src/libraries/Native/Unix/System.Native/browser/pal_random_browser.h:
```diff
@@ -0,0 +1,9 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
I am not sure whether this header file is needed. I think it would be ok to have the function declaration in the .c file.

If you would like to keep it, it would be nice for it to have same name as the `.js` file.

kjpou1 replied:
Not needed and addressed in recent commits

jkotas commented on src/libraries/Native/Unix/System.Native/browser/pal_entropy_bridge.js:
```diff
@@ -0,0 +1,23 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
We do not have other OS-specific System.Native implementation in subdirectories. For example, `pal_log.m` is Apple OS specific and it does not live in subdirectory.

It would be nice to maintain same style.

kjpou1 replied:
renamed and addressed in recent commits

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-crypto-getRandomValues

kjpou1 committed:
Address review comments about directory structure and naming

jkotas commented on src/libraries/Native/Unix/System.Native/pal_random.js:
```diff
@@ -0,0 +1,23 @@
+            return 0;
+        } else {
+            // we couldn't find a proper implementation, as Math.random() is not suitable
+            abort("no cryptographic support found for getRandomValues. Consider polyfilling it if you want to use something insecure like Math.random(), e.g. put this in a --pre-js: var crypto = { getRandomValues: function(array) { for (var i = 0; i < array.length; i++) array[i] = (Math.random()*256)|0 } };");
```
I do not think we should be suggesting an unsecure polyfill in the error message. 

I think we should leave it for the documentation to discuss the polyfill workarounds. The recommended workaround should be something that tries to be secure as much as possible, like the MSR library discussed in the other PR. If somebody wants to use insecure polyfill, I would leave it up to them to figure out how to do it. 

cc @GrabYourPitchforks

kjpou1 replied:
Changed to let normal managed exception to handle error.

```
WASM-ERR: 
WASM-ERR: Unhandled Exception:
WASM-ERR: System.Security.Cryptography.CryptographicException: Error occurred during a cryptographic operation.
WASM-ERR:    at Interop.GetCryptographicallySecureRandomBytes(Byte* buffer, Int32 length)
WASM-ERR:    at System.Security.Cryptography.RandomNumberGeneratorImplementation.GetBytes(Byte* pbBuffer, Int32 count)
WASM-ERR:    at System.Security.Cryptography.RandomNumberGeneratorImplementation.GetBytes(Span`1 data)
WASM-ERR:    at System.Security.Cryptography.RandomNumberGeneratorImplementation.GetNonZeroBytes(Span`1 data)
WASM-ERR:    at System.Security.Cryptography.RandomNumberGeneratorImplementation.GetNonZeroBytes(Byte[] data)

```

Does this work?

jkotas commented on src/mono/wasm/runtime-test.js:
```diff
@@ -62,12 +62,13 @@ if (typeof console !== "undefined") {
 	// /dev/random doesn't work on js shells, so define our own
 	// See library_fs.js:createDefaultDevices ()
 	var crypto = {
 		getRandomValues: function (buffer) {
```
This should have a comment that this is a simple insecure polyfill for testing purposes only.

kjpou1 replied:
Addressed

CoffeeFlux commented on src/mono/wasm/runtime-test.js:
```diff
@@ -62,12 +62,13 @@ if (typeof console !== "undefined") {
 }
 
-if (typeof crypto == 'undefined') {
+if (typeof crypto === 'undefined') {
```
I don't think this is a functional change since typeof should always return a string. Should we be using `===` throughout our JS? I've seen a mix checked in, but if we want to default to the versions without type conversion I'll keep an eye out for that.

lewing replied:
We should clean up these checks in general the match the js version we require but this is fine.

CoffeeFlux commented on src/mono/wasm/runtime-test.js:
```diff
@@ -166,7 +167,7 @@ while (args !== undefined && args.length > 0) {
 		var arg = args [0].substring ("--working-dir=".length);
 		working_dir = arg;
-		args = args.slice (1);
+        args = args.slice (1);
```
I believe this was correct to use tabs before.

kjpou1 replied:
addressed

CoffeeFlux commented on src/mono/wasm/runtime-test.js:
```diff
@@ -62,12 +62,13 @@ if (typeof console !== "undefined") {
 		getRandomValues: function (buffer) {
-			buffer[0] = (Math.random()*256)|0;
+			for (var i = 0; i < buffer.length; i++)
+				buffer[i] = (Math.random()*256)|0;
```
```suggestion
				buffer [i] = (Math.random () * 256) | 0;
```

CoffeeFlux reviewed (commented):
LGTM modulo the things Jan mentioned. Thanks!

marek-safar commented on src/libraries/Native/Unix/System.Native/pal_random.c:
```diff
@@ -63,11 +63,24 @@ Return 0 on success, -1 on failure.
+
+#ifdef __EMSCRIPTEN__
+    extern int32_t dotnet_browser_entropy(uint8_t* buffer, int32_t bufferLength);
+    static bool sMissingBrowserCrypto;
```
I think the caching is redundant

kjpou1 replied:
It follows what is already in place and keeps us from calling external routine again unless I am missing something.  When we were aborting from the js code I would agree but recent mod now returns the -1 from the routine which will allow the managed exception be called.

Let me know if this needs to changed.

marek-safar commented on src/libraries/Native/Unix/System.Native/pal_random.c:
```diff
@@ -63,11 +63,24 @@ Return 0 on success, -1 on failure.
+        else
+            return 0;
+    }
+#else
```
You could also just `return -1` without #else/#endif

kjpou1 replied:
#else was recommended above in previous comment.

lewing reviewed (commented):
Address the open reviews and this looks good.

kjpou1 committed:
Update src/mono/wasm/runtime-test.js

Co-authored-by: Ryan Lucia <ryan@luciaonline.net>

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-crypto-getRandomValues

kjpou1 committed:
Add note about insecure code for testing purposes

kjpou1 committed:
Formatting

kjpou1 committed:
Return -1 if crypto does not exist instead of aborting from js.  This allows the managed code exception flow to continue as normal.
